### PR TITLE
Initial shared memory allocation update

### PR DESCRIPTION
**Pull request by shmsong**
This PR updates the shared memory allocation. Changes include:

- [X] Always allocate shared memory tensors dynamically.
- [X] Always align shared memory allocations to 128b/16B.
- [X] Check available shared memory space for dynamic allocation through driver API calls on compiled kernels, which will account for space already taken statically, e.g. nvfuser_zero.
- [X] Increase shared memory partition if requested dynamic shared memory size exceeds configured limit.
- [X] Assert on requested shared memory size being within the device total shared memory limit.

### CODE REVIEW

File: aten/src/ATen/cuda/detail/LazyNVRTC.cpp [hunk 1]
```diff
@@ -166,7 +166,9 @@ CUDA_STUB1(cuModuleUnload, CUmodule);
 CUDA_STUB3(cuDevicePrimaryCtxGetState, CUdevice, unsigned int *, int *);
 CUDA_STUB4(cuLinkCreate, unsigned int, CUjit_option *, void **, CUlinkState *);
 CUDA_STUB3(cuLinkComplete, CUlinkState, void **, size_t *);
-
+CUDA_STUB3(cuFuncSetAttribute,  CUfunction, CUfunction_attribute, int);
+CUDA_STUB3(cuFuncGetAttribute,int*, CUfunction_attribute, CUfunction);
```
Comment (csarofeen) [inline aten/src/ATen/cuda/detail/LazyNVRTC.cpp]: Nit: formatting

File: test/cpp/jit/test_gpu.cpp [hunk 2]
```diff
@@ -21305,6 +21305,80 @@ TEST_F(NVFuserTest, FusionDoubleBufferVector_CUDA) {
   testValidate(&fusion, cg_outputs, {t0}, {ref}, __LINE__, __FILE__);
 }
 
+// Request 48KB of data in shared mem,
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp:21167]: greater than 48KB and <=64KB is supported on all device types.

File: test/cpp/jit/test_gpu.cpp [hunk 3]
```diff
@@ -21305,6 +21305,80 @@ TEST_F(NVFuserTest, FusionDoubleBufferVector_CUDA) {
...
+  int device_limit = properties->sharedMemPerBlockOptin;
+
+  auto tv0 = makeContigTensor(1);
+  fusion.addInput(tv0);
+  auto tv1 = add(tv0, IrBuilder::create<Double>(1.0));
+  auto tv2 = add(tv1, IrBuilder::create<Double>(2.0));
+  fusion.addOutput(tv2);
+
+  // 4 byte per float
+  tv2->split(0, device_limit / 4);
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp:21218]: Shouldn't this fit? Why can't we hit the exact limit?

File: torch/csrc/jit/codegen/cuda/codegen.cpp [hunk 4]
```diff
@@ -1303,34 +1303,17 @@ class CudaKernelGenerator : private OptOutConstDispatch {
...
-                     << " = reinterpret_cast<" << buffer_dtype << "*>"
-                     << "(array + offset);\n";
-            // Increment Offset Position
-            indent() << "offset += (" << genInline(size) << " * sizeof("
-                     << buffer_dtype << "));\n";
-          }
+          // Align Offset Position
+          indent() << "offset = alignBufferSize(offset,"
+                   // Always align to 128b / 16B
+                   << 16 << ");\n";
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/codegen.cpp:1765]: This seems right to me, but this doesn't seem to agree: https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/executor.cpp#L330-L334

Can you make sure the dynamic shared memory calculation is correct taking into consideration alignment. We may just want to always add 16B per computed allocation in the above executor.cpp code for simplicity.

File: torch/csrc/jit/codegen/cuda/executor.cpp [hunk 5]
```diff
@@ -206,13 +207,16 @@ void FusionExecutor::compileFusion(
 
   const auto& kernel_summary = kernel->summary();
 
+  // We currently shouldn't allocate any more shared mem
+  //  tensors statically but could keep this path if
+  //  needed in later development.
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/executor.cpp:212]: Please remove the static paths in a follow up. The other thing we may want to do is also get rid of the second expression evaluator in a cleanup PR since `kir::Kernel` inherits from `Fusion` now.

File: torch/csrc/jit/codegen/cuda/codegen.cpp (hunk 4 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/codegen.cpp:1765]: It would be good to explicitly test the alignment, but I can't think of a good way to get a deterministic ordering of shared memory allocations so that the first will throw the alignment off for the second and we try to do vectorized loads with the second.

File: torch/csrc/jit/codegen/cuda/executor.cpp [hunk 6]
```diff
@@ -529,15 +541,26 @@ LaunchParams FusionExecutor::computeLaunchParams(
...
-      max_device_smem);
+  // Check that requested smem size can be dynamically allocated.
+  //  This check is only done once a kernel has been compiled, since
+  //  maybe_available_dynamic_smem_ needs to be evaluated on
+  //  a compiled kernel.
+  if (maybe_available_dynamic_smem_.has_value()) {
+    maybe_available_dynamic_smem_ = maybe_available_dynamic_smem_.value();
+    TORCH_INTERNAL_ASSERT(
+        (dynamic_smem_size) < (maybe_available_dynamic_smem_.value() +
+                               device_smem_limit_ - configured_device_smem_),
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/executor.cpp]: This seems a little convoluted to me, why isn't this not just `dynamic_smem_size < device_smem_limit_` ?

**Review by csarofeen — COMMENTED** [review]
The only required comment I'd like to see addressed before merging this PR is to check the dynamic shared memory calculation under alignment: https://github.com/csarofeen/pytorch/pull/1498#discussion_r825325088

Everything else is a suggestion, or cleanup that can be done in later PRs.

File: aten/src/ATen/cuda/detail/LazyNVRTC.cpp (hunk 1 above)
Comment (shmsong) [reply aten/src/ATen/cuda/detail/LazyNVRTC.cpp]: Updated. Thanks!

File: test/cpp/jit/test_gpu.cpp (hunk 2 above)
Comment (shmsong) [reply test/cpp/jit/test_gpu.cpp:21167]: Yes and I think 48KB has been the boundary and dynamic allocation is required beyond this. This test case is testing that the fusion tensors are static shaped but the size would require dynamic allocation.

File: test/cpp/jit/test_gpu.cpp (hunk 3 above)
Comment (shmsong) [reply test/cpp/jit/test_gpu.cpp:21218]: Currently due to magic zero taking an integer on sharedmem.

This test case just wanted to make sure we throw at exactly the boundary and it will fail once we eventually don't need magic zero anymore so we could update the new boundary value.

File: torch/csrc/jit/codegen/cuda/codegen.cpp (hunk 4 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/codegen.cpp:1765]: I think it is correct since the no alignment padding case in executor.cpp is only used for static smems. It looks like we want to remove the static allocation path so I will just remove that part as well in the same follow up.

Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/codegen.cpp:1765]: It'd be great idea to add an alignment test. I could try to do a fusion with just a few tensors with different shapes and see if it covers the code paths.

File: torch/csrc/jit/codegen/cuda/executor.cpp (hunk 5 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/executor.cpp:212]: Makes sense. Will do in a follow up. Thanks.

File: torch/csrc/jit/codegen/cuda/executor.cpp (hunk 6 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/executor.cpp]: The extra complication comes from `maybe_available_dynamic_smem_`, and `configured_device_smem_`, both from driver api. 

basically we want to supply `dynamic_smem_size` with two possible components of memory space:

`maybe_available_dynamic_smem_` = amount of memory we still have in `configured_device_smem_` without re-configuring smem (static sharedmem allocs, aka. magic zero only at this point, already subtracted).

and

`device_smem_limit_ - configured_device_smem_` = max additional shared memory we can request from L1 through re-configuration.

File: torch/csrc/jit/codegen/cuda/codegen.cpp (hunk 4 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/codegen.cpp:1765]: I see. Sorry I missed the alignment size on the last pass. Just updated so that executor always align to 128b as well.

**Review by csarofeen — APPROVED** [review]
LGTM thanks for the updates.

File: torch/csrc/jit/codegen/cuda/executor.cpp (hunk 6 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/executor.cpp]: Maybe it would be more clear if you describe all variables in a comment, maybe just `maybe_available_dynamic_smem`.

Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/executor.cpp]: renamed and added comments. Thanks.
